# Выбор региона добычи нефти с наибольшей прибылью

В добывающей компании «ГлавРосГосНефть» нужно решить, где бурить новую скважину.
Шаги для выбора локации обычно такие:<br />
•	В избранном регионе собирают характеристики для скважин: качество нефти и объём её запасов;<br />
•	Строят модель для предсказания объёма запасов в новых скважинах;<br />
•	Выбирают скважины с самыми высокими оценками значений;<br />
•	Определяют регион с максимальной суммарной прибылью отобранных скважин.<br />
Предоставлены пробы нефти в трёх регионах. Характеристики для каждой скважины в регионе уже известны. Требуется построить модель для определения региона, где добыча принесёт наибольшую прибыль. Необходимо проанализировать возможную прибыль и риски техникой Bootstrap.

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Введение" data-toc-modified-id="Введение-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Введение</a></span><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Условия-задачи" data-toc-modified-id="Условия-задачи-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Условия задачи</a></span></li></ul></li><li><span><a href="#Считывание-данных-и-импорт-библиотек" data-toc-modified-id="Считывание-данных-и-импорт-библиотек-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Считывание данных и импорт библиотек</a></span><ul class="toc-item"><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Считывание-и-просмотр-данных" data-toc-modified-id="Считывание-и-просмотр-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Считывание и просмотр данных</a></span></li><li><span><a href="#Проверка-на-пустые-значения-и-дубликаты." data-toc-modified-id="Проверка-на-пустые-значения-и-дубликаты.-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Проверка на пустые значения и дубликаты.</a></span></li></ul></li><li><span><a href="#Обучение-моделей-для-каждого-региона-и-расчёт-RMSE-модели" data-toc-modified-id="Обучение-моделей-для-каждого-региона-и-расчёт-RMSE-модели-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Обучение моделей для каждого региона и расчёт RMSE модели</a></span><ul class="toc-item"><li><span><a href="#Функция-обучение-модели,-получение-предсказаний,-расчёт-среднего-запаса-сырья-и-RMSE" data-toc-modified-id="Функция-обучение-модели,-получение-предсказаний,-расчёт-среднего-запаса-сырья-и-RMSE-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Функция обучение модели, получение предсказаний, расчёт среднего запаса сырья и RMSE</a></span></li><li><span><a href="#Регион-0" data-toc-modified-id="Регион-0-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Регион 0</a></span></li><li><span><a href="#Регион-1" data-toc-modified-id="Регион-1-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Регион 1</a></span></li><li><span><a href="#Регион-2" data-toc-modified-id="Регион-2-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Регион 2</a></span></li><li><span><a href="#Сравнительная-таблица-средних-значений-и-RMSE-по-регионам" data-toc-modified-id="Сравнительная-таблица-средних-значений-и-RMSE-по-регионам-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Сравнительная таблица средних значений и RMSE по регионам</a></span></li></ul></li><li><span><a href="#Подготовка-к-расчёту-прибыли" data-toc-modified-id="Подготовка-к-расчёту-прибыли-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Подготовка к расчёту прибыли</a></span><ul class="toc-item"><li><span><a href="#Ключевые-значения-для-расчётов" data-toc-modified-id="Ключевые-значения-для-расчётов-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Ключевые значения для расчётов</a></span></li><li><span><a href="#Достаточный-объём-сырья-для-безубыточной-разработки-скважины" data-toc-modified-id="Достаточный-объём-сырья-для-безубыточной-разработки-скважины-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Достаточный объём сырья для безубыточной разработки скважины</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Расчёт-рисков-и-прибыли-для-каждого-региона-с-использованием-метода-bootstrap" data-toc-modified-id="Расчёт-рисков-и-прибыли-для-каждого-региона-с-использованием-метода-bootstrap-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Расчёт рисков и прибыли для каждого региона с использованием метода bootstrap</a></span><ul class="toc-item"><li><span><a href="#Функция-расчёта-прибыли" data-toc-modified-id="Функция-расчёта-прибыли-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Функция расчёта прибыли</a></span></li><li><span><a href="#Функция-расчёта-распределения-прибыли-методом-bootstrap" data-toc-modified-id="Функция-расчёта-распределения-прибыли-методом-bootstrap-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Функция расчёта распределения прибыли методом bootstrap</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

## Введение

### Описание данных

Данные геологоразведки трёх регионов находятся в файлах: geo_data_0.csv, geo_data_1.csv, geo_data_2.csv

- id — уникальный идентификатор скважины;
- f0, f1, f2 — три признака точек (неважно, что они означают, но сами признаки значимы);
- product — объём запасов в скважине (тыс. баррелей).

### Условия задачи

- Для обучения модели подходит только линейная регрессия (остальные — недостаточно предсказуемые).
- При разведке региона исследуют 500 точек, из которых с помощью машинного обучения выбирают 200 лучших для разработки.
- Бюджет на разработку скважин в регионе — 10 млрд рублей.
- При нынешних ценах один баррель сырья приносит 450 рублей дохода. Доход с каждой единицы продукта составляет 450 тыс. рублей,  поскольку объём указан в тысячах баррелей.
- После оценки рисков нужно оставить лишь те регионы, в которых вероятность убытков меньше 2.5%. Среди них выбирают регион с наибольшей средней прибылью. <br>


Данные синтетические: детали контрактов и характеристики месторождений не разглашаются.

## Считывание данных и импорт библиотек

### Импорт библиотек

In [1]:
import pandas as pd
import statistics as st
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt
import numpy as np

### Считывание и просмотр данных

In [2]:
try:
    data_0 = pd.read_csv('/datasets/geo_data_0.csv')
    data_1 = pd.read_csv('/datasets/geo_data_1.csv')
    data_2 = pd.read_csv('/datasets/geo_data_2.csv')
except:
    data_0 = pd.read_csv('datasets/geo_data_0.csv')
    data_1 = pd.read_csv('datasets/geo_data_1.csv')
    data_2 = pd.read_csv('datasets/geo_data_2.csv')

In [3]:
data_0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [4]:
data_0.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [5]:
data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [6]:
data_1.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [7]:
data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [8]:
data_2.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

### Проверка на пустые значения и дубликаты.

In [9]:
data_0.isna().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

In [10]:
data_0.duplicated().sum()

0

Пустых значений и явных дубликатов в данных региона 0 не обнаружено.

In [11]:
data_1.isna().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

In [12]:
data_1.duplicated().sum()

0

Пустых значений и явных дубликатов в данных региона 1 не обнаружено.

In [13]:
data_2.isna().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

In [14]:
data_2.duplicated().sum()

0

Пустых значений и явных дубликатов в данных региона 2 не обнаружено.

Проверка дубликатов в столбце id.

In [15]:
data_0['id'].duplicated().sum()

10

In [16]:
data_1['id'].duplicated().sum()

4

In [17]:
data_2['id'].duplicated().sum()

4

Обнаружено 4, 10 дубликатов. Это составляет 0,0001 долю от всех значений. Это очень маленькое значение, которое не окажет какого-либо влияния на расчёты. Поэтому не будем удалять эти значения.

## Обучение моделей для каждого региона и расчёт RMSE модели

### Функция обучение модели, получение предсказаний, расчёт среднего запаса сырья и RMSE

In [18]:
# Значение для передачи в аргумент random_state
state = 94837

Так как обработка данных с помощью машинного обучения будет совпадать для всех 3-х регионов создадим для этого функцию. Функция будет принимать на вход данные скважин региона, разделять данные, обучать модель и выдавать рассчитанные данные: значения целевого показателя для валидационной выборки, предсказания для валидационной выборки, среднее значение предсказаний и значение RMSE.

In [19]:
# Функция обучение модели, получение предсказаний, расчёт среднего запаса сырья и RMSE
def lr_rmse (data):
    features= data.drop(columns='product')
    target = data['product']
    features_train, features_valid, target_train, target_valid = train_test_split(
        features, target, test_size=0.25, random_state=state)
    model = LinearRegression()
    model.fit(features_train, target_train)
    prediction = model.predict(features_valid)
    prediction_mean = st.mean(prediction)
    prediction = pd.Series(prediction) # создание Series из предсказаний)
    target_valid = target_valid.reset_index(drop=True) # сброс индексов target_valid, чтобы они совпадали с индексами prediction
    rmse = (mean_squared_error(target_valid, prediction)) ** 0.5
    return(target_valid, prediction, prediction_mean, rmse)

### Регион 0

In [20]:
data_0 = data_0.drop(columns='id')

In [21]:
target_valid_0, prediction_0, prediction_mean_0, rmse_0 = lr_rmse(data_0)
print('Среднее предсказанное значение добычи с 1 скважины в 0-ом регионе', round(prediction_mean_0, 2), 'тыс баррелей, среднеквадратичная ошибка', round(rmse_0, 2), 'тыс баррелей')

Среднее предсказанное значение добычи с 1 скважины в 0-ом регионе 92.57 тыс баррелей, среднеквадратичная ошибка 37.43 тыс баррелей


### Регион 1

In [22]:
data_1 = data_1.drop(columns='id')

In [23]:
target_valid_1, prediction_1, prediction_mean_1, rmse_1 = lr_rmse(data_1)
print('Среднее предсказанное значение добычи с 1 скважины в 1-ом регионе', round(prediction_mean_1, 2), 'тыс баррелей, среднеквадратичная ошибка', round(rmse_1, 2), 'тыс баррелей')

Среднее предсказанное значение добычи с 1 скважины в 1-ом регионе 68.59 тыс баррелей, среднеквадратичная ошибка 0.89 тыс баррелей


### Регион 2

In [24]:
data_2 = data_2.drop(columns='id')

In [25]:
target_valid_2, prediction_2, prediction_mean_2, rmse_2 = lr_rmse(data_2)
print('Среднее предсказанное значение добычи с 1 скважины во 2-ом регионе', round(prediction_mean_2, 2), 'тыс баррелей, среднеквадратичная ошибка', round(rmse_2, 2), 'тыс баррелей')

Среднее предсказанное значение добычи с 1 скважины во 2-ом регионе 94.93 тыс баррелей, среднеквадратичная ошибка 40.24 тыс баррелей


### Сравнительная таблица средних значений и RMSE по регионам

In [26]:
table = {
    'region': [0, 1,2],
    'prediction_mean' : [prediction_mean_0, prediction_mean_1, prediction_mean_2],
    'rmse' : [rmse_0, rmse_1, rmse_2]
        }
region_prediction = pd.DataFrame(table)
region_prediction['prediction_mean'] = round(region_prediction['prediction_mean'], 2)
region_prediction['rmse'] = round(region_prediction['rmse'], 2)
display(region_prediction)

region  prediction_mean   rmse
0       0            92.57  37.43
1       1            68.59   0.89
2       2            94.93  40.24

Из полученной таблицы видно, что среднее значение предсказанного объёма на 1 скважину больше в регионе 0 и 2, по сравнению с регионом 1. В то же время, разброс значений в регионе 1 значительно ниже, чем в регионе 0 и 2. 

## Подготовка к расчёту прибыли

### Ключевые значения для расчётов

In [27]:
WELLS_EXPLORATION = 500 #количество разведочных скважин
WELLS_DEVELOPMENT = 200 # количество скважин для разработки
DEVELOPMENT_BUDGET = 10000000000 # бюджет на разработку скважин в регионе
PRODUCT_REVENUE = 450000 # доход с каждой единицы продукта

### Достаточный объём сырья для безубыточной разработки скважины

In [28]:
volume_break_even = DEVELOPMENT_BUDGET / (PRODUCT_REVENUE * WELLS_DEVELOPMENT)
print('Более', round(volume_break_even, 2), 'тыс барелей необходимо добывать в среднем на одну скважину, чтобы добыча была прибыльной.')

Более 111.11 тыс барелей необходимо добывать в среднем на одну скважину, чтобы добыча была прибыльной.


### Вывод

Значение достаточного сырья на 1 скважину 111.11 тыс баррелей, больше средних запасов предсказанного сырья на 1 скважину для всех регионов.

## Расчёт рисков и прибыли для каждого региона с использованием метода bootstrap

### Функция расчёта прибыли

In [29]:
def profit (target_valid, prediction):
    prediction_200 = prediction.sort_values(ascending=False)[:WELLS_DEVELOPMENT] # отбор 200 скважин с самыми большими запасами
    profit = sum(target_valid[prediction_200.index]) *  PRODUCT_REVENUE - DEVELOPMENT_BUDGET
    return(profit)

### Функция расчёта распределения прибыли методом bootstrap

Так как расчёт распределения прибыли будет проводится для всех регионов сделаем для этого функцию. На вход функция будет получать целевые и предсказаные значения валидационной выборки. В теле функции будет произвдён расчёт значений прибыли для разных подбвыборок полученных методом bootstrap с возвращающимися значениями. На выходе функции будут среднее значение прибыли, нижнее и верхнее значение 95% квантиля и вероятность убытков (отношение количества уботочных (отрицательных) значений прибыли к общему количеству выборок)

In [30]:
number_bootstrap = 1000 # количество быборок для bootstrap
state_b = np.random.RandomState(57865)
def bootstrap (target_valid, prediction):
    profit_data = []
    for i in range(number_bootstrap):
        prediction_subsample = prediction.sample(frac=WELLS_EXPLORATION / len(prediction), replace=True, random_state=state_b)
        profit_data.append(profit(target_valid, prediction_subsample))
    profit_mean = np.mean(profit_data)
    profit_data = pd.Series(profit_data)
    profit_lower = profit_data.quantile(0.025)
    profit_upper = profit_data.quantile(0.975)
    losses = profit_data[profit_data < 0].count() / number_bootstrap
    return([profit_mean, profit_lower, profit_upper, losses])

In [31]:
region_0 = bootstrap(target_valid_0, prediction_0)
print(region_0)

[420399445.66394556, -75225279.25756153, 924777302.5141213, 0.041]


In [32]:
region_1 = bootstrap(target_valid_1, prediction_1)
print(region_1)

[401153939.6315656, 20302930.659739595, 800895286.7302871, 0.021]


In [33]:
region_2 = bootstrap(target_valid_2, prediction_2)
print(region_2)

[348442344.43127066, -132508549.476716, 844773614.6541469, 0.089]


In [34]:
table_total = pd.DataFrame(columns=['profit_mean_milliard', 'profit_2.5%', 'profit_97.5%', 'losses_%'])
table_total.loc[0] = region_0
table_total.loc[1] = region_1
table_total.loc[2] = region_2
table_total.insert(0, 'region', [0, 1, 2])
table_total['profit_mean_milliard'] = round(table_total['profit_mean_milliard']/1000000000, 2)
table_total['profit_2.5%'] = round(table_total['profit_2.5%']/1000000000, 2)
table_total['profit_97.5%'] = round(table_total['profit_97.5%']/1000000000, 2)
table_total['losses_%'] = round(table_total['losses_%']*100, 1)
display(table_total)

region  profit_mean_milliard  profit_2.5%  profit_97.5%  losses_%
0       0                  0.42        -0.08          0.92       4.1
1       1                  0.40         0.02          0.80       2.1
2       2                  0.35        -0.13          0.84       8.9

## Вывод

Из полученной таблицы видно, что в среднем во всех регионах добыча принесёт прибыль, есть только разница в рисках получения убытков. Самый маленький риск получения убытков 2,1% в регионе 2. Таким образом для добычи рекомендуется 2-ой регион.